internal/pkg/override/patch.go

```diff
@@ -0,0 +1,283 @@
+	if err := yaml.Unmarshal(body, &root); err != nil {
+		return nil, fmt.Errorf("invalid template: %w", err)
+	}
+
+	for _, patch := range patches {
+		var err error
+		switch patch.Operation {
+		case "add":
+			err = patch.applyAdd(&root)
+		case "remove":
+			err = patch.applyRemove(&root)
+		case "replace":
+			err = patch.applyReplace(&root)
+		default:
+			return nil, fmt.Errorf("unsupported operation %q: supported operations are %q, %q, and %q.", patch.Operation, "add", "remove", "replace")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("unable to apply %q patch: %w", patch.Operation, err)
+		}
+	}
+
+	out, err := yaml.Marshal(&root)
+	if err != nil {
+		return nil, fmt.Errorf("unable to return modified document to []byte: %w", err)
+	}
+	return out, nil
+}
+
+func unmarshalPatches(path string, fs afero.Fs) ([]yamlPatch, error) {
+	content, err := afero.ReadFile(fs, path)
+	if err != nil {
+		return nil, fmt.Errorf("read file at %q: %w", path, err)
+	}
+
+	var patches []yamlPatch
+	if err := yaml.Unmarshal(content, &patches); err != nil {
+		return nil, fmt.Errorf("file at %q does not conform to the YAML patch document schema: %w", path, err)
+	}
+
+	return patches, nil
+}
+
+func (p *yamlPatch) applyAdd(root *yaml.Node) error {
+	if p.Value.IsZero() {
+		return fmt.Errorf("value required")
+	}
+
+	return followJSONPointer(root, p.Path, func(node *yaml.Node, pointer []string) error {
+		if len(pointer) != 1 {
+			return nil
+		}
+
+		switch node.Kind {
+		case yaml.DocumentNode:
+			return p.encodeAndStop(node)
+		case yaml.MappingNode:
+			// if the key is in this map, they are trying to replace it
+			for i := 0; i < len(node.Content); i += 2 {
+				if node.Content[i].Value == pointer[0] {
+					return p.encodeAndStop(node.Content[i+1])
+				}
+			}
+
+			// if the key isn't in this map, then we need to create it for them
+			node.Content = append(node.Content, &yaml.Node{
+				Kind:  yaml.ScalarNode,
+				Tag:   "!!str",
+				Value: pointer[0],
+			})
+			node.Content = append(node.Content, &p.Value)
+			return errStopFollowingPointer
+		case yaml.SequenceNode:
+			if pointer[0] == "-" {
+				// add to end of sequence
+				node.Content = append(node.Content, &p.Value)
+				return errStopFollowingPointer
+			}
+
+			idx, err := strconv.Atoi(pointer[0])
+			switch {
+			case err != nil:
+				return fmt.Errorf("expected index in sequence, got %q", pointer[0])
+			case idx < 0 || idx > len(node.Content):
+				return fmt.Errorf("invalid index %d for sequence of length %d", idx, len(node.Content))
+			}
+
+			// add node at idx
+			node.Content = append(node.Content[:idx], append([]*yaml.Node{&p.Value}, node.Content[idx:]...)...)
+			return errStopFollowingPointer
+		}
+
+		return nil
+	})
+}
+
+func (p *yamlPatch) applyRemove(root *yaml.Node) error {
+	return followJSONPointer(root, p.Path, func(node *yaml.Node, pointer []string) error {
+		if len(pointer) != 1 {
+			return nil
+		}
+
+		switch node.Kind {
+		case yaml.DocumentNode:
```

Comment:
How would we hit this case?

Reply:
it's a fairly...useless scenario for our use case, but with the following patch:
```
op: remove
path: ""
```
(see this test https://github.com/aws/copilot-cli/pull/4641/files#diff-db45db57a8c5fa7983bfd1bd51464cb707a656831be2aa65c72bf3f758cf296cR344-R350)